From bfd03698e85a6bff663280e538dab2eea4fb90ff Mon Sep 17 00:00:00 2001
From: Calvin Johnson <calvin.johnson@nxp.com>
Date: Wed, 20 Jun 2018 10:23:01 +0530
Subject: [PATCH 19/22] staging: fsl_ppfe/eth: define pfe ndo_change_mtu
 function

Define ndo_change_mtu function for pfe. This sets the max Rx frame
length to the new mtu.

Signed-off-by: Calvin Johnson <calvin.johnson@nxp.com>

[ Srinidhi Rao : Ported this patch to photon linux from
  qoriq-components linux 'linux-4.14-nxp'
  (https://source.codeaurora.org/external/qoriq/qoriq-components/linux)
]

Signed-off-by: srinidhira0 <srinidhir@vmware.com>
---
 drivers/staging/fsl_ppfe/include/pfe/pfe.h |  1 +
 drivers/staging/fsl_ppfe/pfe_eth.c         | 12 ++++++++++++
 drivers/staging/fsl_ppfe/pfe_hal.c         | 11 +++++++++++
 3 files changed, 24 insertions(+)

diff --git a/drivers/staging/fsl_ppfe/include/pfe/pfe.h b/drivers/staging/fsl_ppfe/include/pfe/pfe.h
index d93ae4c60251..79eeafa60eb6 100644
--- a/drivers/staging/fsl_ppfe/include/pfe/pfe.h
+++ b/drivers/staging/fsl_ppfe/include/pfe/pfe.h
@@ -303,6 +303,7 @@ void gemac_allow_broadcast(void *base);
 void gemac_no_broadcast(void *base);
 void gemac_enable_1536_rx(void *base);
 void gemac_disable_1536_rx(void *base);
+void gemac_set_rx_max_fl(void *base, int mtu);
 void gemac_enable_rx_jmb(void *base);
 void gemac_disable_rx_jmb(void *base);
 void gemac_enable_stacked_vlan(void *base);
diff --git a/drivers/staging/fsl_ppfe/pfe_eth.c b/drivers/staging/fsl_ppfe/pfe_eth.c
index 0aa08d251b7c..77d7a7b682c6 100644
--- a/drivers/staging/fsl_ppfe/pfe_eth.c
+++ b/drivers/staging/fsl_ppfe/pfe_eth.c
@@ -1345,6 +1345,17 @@ static int pfe_eth_event_handler(void *data, int event, int qno)
 	return 0;
 }
 
+static int pfe_eth_change_mtu(struct net_device *ndev, int new_mtu)
+{
+	struct pfe_eth_priv_s *priv = netdev_priv(ndev);
+
+	ndev->mtu = new_mtu;
+	new_mtu += ETH_HLEN + ETH_FCS_LEN;
+	gemac_set_rx_max_fl(priv->EMAC_baseaddr, new_mtu);
+
+	return 0;
+}
+
 /* pfe_eth_open
  */
 static int pfe_eth_open(struct net_device *ndev)
@@ -2246,6 +2257,7 @@ static const struct net_device_ops pfe_netdev_ops = {
 	.ndo_set_rx_mode = pfe_eth_set_multi,
 	.ndo_set_mac_address = pfe_eth_set_mac_address,
 	.ndo_validate_addr = eth_validate_addr,
+	.ndo_change_mtu = pfe_eth_change_mtu,
 	.ndo_get_stats = pfe_eth_get_stats,
 	.ndo_set_features = pfe_eth_set_features,
 };
diff --git a/drivers/staging/fsl_ppfe/pfe_hal.c b/drivers/staging/fsl_ppfe/pfe_hal.c
index 7e89606bb40c..cb8cb9cbd42f 100644
--- a/drivers/staging/fsl_ppfe/pfe_hal.c
+++ b/drivers/staging/fsl_ppfe/pfe_hal.c
@@ -1017,6 +1017,17 @@ void gemac_enable_1536_rx(void *base)
 		| (1536 << 16), base +	EMAC_RCNTRL_REG);
 }
 
+/* GEMAC set rx Max frame length.
+ * @param[in]	base	GEMAC base address
+ * @param[in]	mtu	new mtu
+ */
+void gemac_set_rx_max_fl(void *base, int mtu)
+{
+	/* Set mtu as Maximum frame length */
+	writel((readl(base + EMAC_RCNTRL_REG) & PFE_RCR_MAX_FL_MASK)
+		| (mtu << 16), base + EMAC_RCNTRL_REG);
+}
+
 /* GEMAC enable jumbo function.
  * @param[in]	base	GEMAC base address
  */
-- 
2.14.2

